homeassistant/components/cover/device_trigger.py

```diff
@@ -0,0 +1,211 @@
+"""Provides device automations for Cover."""
+from typing import List
+import voluptuous as vol
+
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_SUPPORTED_FEATURES,
+    CONF_ABOVE,
+    CONF_BELOW,
+    CONF_DOMAIN,
+    CONF_TYPE,
+    CONF_PLATFORM,
+    CONF_DEVICE_ID,
+    CONF_ENTITY_ID,
+    STATE_CLOSED,
+    STATE_CLOSING,
+    STATE_OPEN,
+    STATE_OPENING,
+)
+from homeassistant.core import HomeAssistant, CALLBACK_TYPE
+from homeassistant.helpers import config_validation as cv, entity_registry
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.components.automation import (
+    state as state_automation,
+    template as template_automation,
+    AutomationActionType,
+)
+from homeassistant.components.device_automation import TRIGGER_BASE_SCHEMA
+from . import (
+    DOMAIN,
+    SUPPORT_CLOSE,
+    SUPPORT_OPEN,
+    SUPPORT_SET_POSITION,
+    SUPPORT_SET_TILT_POSITION,
+)
+
+POSITION_TRIGGER_TYPES = {"position", "tilt_position"}
+STATE_TRIGGER_TYPES = {"opened", "closed", "opening", "closing"}
+
+POSITION_TRIGGER_SCHEMA = vol.All(
+    TRIGGER_BASE_SCHEMA.extend(
+        {
+            vol.Required(CONF_ENTITY_ID): cv.entity_id,
+            vol.Required(CONF_TYPE): vol.In(POSITION_TRIGGER_TYPES),
+            vol.Optional(CONF_ABOVE): vol.All(
+                vol.Coerce(int), vol.Range(min=0, max=100)
+            ),
+            vol.Optional(CONF_BELOW): vol.All(
+                vol.Coerce(int), vol.Range(min=0, max=100)
+            ),
+        }
+    ),
+    cv.has_at_least_one_key(CONF_BELOW, CONF_ABOVE),
+)
+
+STATE_TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_TYPE): vol.In(STATE_TRIGGER_TYPES),
+    }
+)
+
+TRIGGER_SCHEMA = vol.Any(POSITION_TRIGGER_SCHEMA, STATE_TRIGGER_SCHEMA)
+
+
+async def async_get_triggers(hass: HomeAssistant, device_id: str) -> List[dict]:
+    """List device triggers for Cover devices."""
+    registry = await entity_registry.async_get_registry(hass)
+    triggers = []
+
+    # Get all the integrations entities for this device
+    for entry in entity_registry.async_entries_for_device(registry, device_id):
+        if entry.domain != DOMAIN:
+            continue
+
+        state = hass.states.get(entry.entity_id)
+        if not state or ATTR_SUPPORTED_FEATURES not in state.attributes:
+            continue
+
+        supported_features = state.attributes[ATTR_SUPPORTED_FEATURES]
+        supports_open_close = supported_features & (SUPPORT_OPEN | SUPPORT_CLOSE)
+
+        # Add triggers for each entity that belongs to this integration
+        if supports_open_close:
+            triggers.append(
+                {
+                    CONF_PLATFORM: "device",
+                    CONF_DEVICE_ID: device_id,
+                    CONF_DOMAIN: DOMAIN,
+                    CONF_ENTITY_ID: entry.entity_id,
+                    CONF_TYPE: "opened",
```

Comment:
It will be very confusing that we have opened and opening. Is there a way we could combine those things?

Reply:
Hopefully the strings are a little bit clearer now:
```json
"device_automation": {
    "condition_type": {
      "is_open": "{entity_name} is open",
      "is_closed": "{entity_name} is closed",
      "is_opening": "{entity_name} is opening",
      "is_closing": "{entity_name} is closing",
      "is_position": "Current {entity_name} position is",
      "is_tilt_position": "Current {entity_name} tilt position is"
    },
    "trigger_type": {
      "opened": "{entity_name} was opened",
      "closed": "{entity_name} was closed",
      "opening": "{entity_name} starts to open",
      "closing": "{entity_name} starts to close",
      "position": "{entity_name} position changes",
      "tilt_position": "{entity_name} tilt position changes"
    }
```